internal/provider/factory.go

```diff
@@ -68,76 +72,91 @@ type Factory interface {
+func (f *factory) ProviderFor(ctx context.Context, pa v1alpha1.ProviderAccessor, c Config) (provider Provider, err error) {
+	if !pa.IsDelegating() {
+		if pa.GetProviderRef().Name == "" {
+			return nil, errProviderRefRequired
 		}
-	} else {
-		providerSecret = &v1.Secret{
+		pSecret := &v1.Secret{
 			ObjectMeta: metav1.ObjectMeta{
 				Name:      pa.GetProviderRef().Name,
 				Namespace: pa.GetNamespace(),
 			}}
+		if err = f.Client.Get(ctx, client.ObjectKeyFromObject(pSecret), pSecret); err != nil {
+			return
+		}
 
-		if err := f.Client.Get(ctx, client.ObjectKeyFromObject(providerSecret), providerSecret); err != nil {
-			return nil, err
+		provider, err = f.providerForSecret(ctx, pSecret, c)
+		if err != nil {
+			return
 		}
 	}
 
-	provider, err := NameForProviderSecret(providerSecret)
+	if pa.IsDelegating() || requiresDelegation(provider, pa) {
+		return f.delegationProviderFor(ctx, pa, c)
+	}
+
+	return
+}
+
+func (f *factory) delegationProviderFor(ctx context.Context, pa v1alpha1.ProviderAccessor, pConfig Config) (Provider, error) {
+	if f.delegationProviderFunc == nil {
+		return nil, errDelegationProviderNotConfigured
+	}
+	return f.delegationProviderFunc(ctx, f.Client, f.dynamicClient, pa, pConfig)
+}
+
+func (f *factory) providerForSecret(ctx context.Context, pSecret *v1.Secret, pConfig Config) (Provider, error) {
+	logger := log.FromContext(ctx)
+
+	providerName, err := NameForProviderSecret(pSecret)
 	if err != nil {
 		return nil, err
 	}
 
 	constructorsLock.RLock()
 	defer constructorsLock.RUnlock()
-	if constructor, ok := constructors[provider]; ok {
-		if !slices.Contains(f.providers, provider) {
-			return nil, fmt.Errorf("provider '%s' not enabled", provider)
+
+	if constructor, ok := constructors[providerName]; ok {
+		if !slices.Contains(f.providers, providerName) {
+			return nil, fmt.Errorf("provider '%s' not enabled", providerName)
 		}
-		logger.V(1).Info(fmt.Sprintf("initializing %s provider with config", provider), "config", c)
+		logger.V(1).Info(fmt.Sprintf("initializing %s provider with config", providerName), "config", pConfig)
 		switch typedConstructor := constructor.(type) {
 		case ProviderConstructor:
-			return typedConstructor(ctx, providerSecret, c)
+			return typedConstructor(ctx, pSecret, pConfig)
 		case ProviderConstructorWithClient:
-			return typedConstructor(ctx, f.dynamicClient, providerSecret, c)
+			return typedConstructor(ctx, f.dynamicClient, pSecret, pConfig)
 		}
-
-		return nil, fmt.Errorf("unrecognised contructor for provider '%s'", provider)
+		return nil, fmt.Errorf("unrecognised contructor for provider '%s'", providerName)
 	}
+	return nil, fmt.Errorf("providerName '%s' not registered", providerName)
+}
 
-	return nil, fmt.Errorf("provider '%s' not registered", provider)
+// requiresDelegation return true if the given provider requires delegation be enforced for the given resource(DNSRecord)
+func requiresDelegation(_ Provider, _ v1alpha1.ProviderAccessor) bool {
```

Comment:
is this how you expect as in always returning false?

Reply:
When CoreDNS is updated this be changed, for now it just returns false.